### PR TITLE
Tfdataset using map instead of from_generator

### PR DESCRIPTION
**Pull request by frodre**
The using `tf.Dataset.map` to map across netcdf filenames, open them, and convert to tensors allows for use of interleave, which can parallelize data input, and allows for reshuffling the file order with every epoch (not possible w/ generator currently).  

From our memory leak investigation, shuffling across samples incurs a severe input-pipeline performance cost, but is required to randomize our `from_generator` input pipeline.  Based on tensorflow profileing~60% of the time is spent waiting for inputs when sample shuffling is used vs. 30% when no sample shuffling is performed for i/o bound model training.  (Tested with 100 files over 10 epochs)

Using map, we might be able to remove the need for sample shuffling since interleave will mix samples from shuffled file ordering.  Using a draft PR to test that offline performance is not significantly reduced with production-like workflow.

### CODE REVIEW

**Comment by frodre** [comment]
The map operation is quite a bit slower in this context.  Perhaps something in the transformation from xarray -> tensor map takes extra time here?  Regardless, closing for now

image
[wandb](https://wandb.ai/ai2cm/microphysics-emulation/groups/sample-buffer-generator_or_map-mar-2022/workspace?workspace=user-aperkins)
